Add loopUntilNoMoreEvents() API to wait for arbitrary code.

achimnol opened the pull request:
- Existing APIs rely on explicit callback argument or predicate.
  This API does not require such prior information, just like
  how the nodejs main loop continues or terminates.
- To use this API to block for a specific region of code, you should
  take care of other existing callbacks using unref() and ref() methods
  to prevent them from making loopUntilNoEvents() to block "longer"
  than you want.

abbr commented:
Can you give a use case where the new api is needed? I don't see adding a predicate as inconvenient even when result of async call is not a concern but only the timing matters.
Also, I don't get how can the new api prevent blocking longer than necessary. I suspect the api will be more prone to deadlock. Can you elaborate your 2nd bullet point?

achimnol commented:
This is required to execute arbitrary user-input codes, using `eval()` or `vm.Script` objects, where I cannot determine which callback is the "last" executed one and augment the user code accordingly in a reliable way.
Of course, this is very "dangerous" for normal node users and I even won't recommend to use this except for very specific use cases.

achimnol committed:
Added loopUntilNoMoreEvents() API to wait for aribtrary code.

 * Existing APIs rely on explicit callback argument or predicate.
   This API does not require such prior information, just like
   how the node.js main loop continues or terminates.

 * To use this API to block for a specific region of code, you should
   take care of other existing callbacks using unref() and ref() methods
   to prevent them from making loopUntilNoMoreEvents() to block "longer"
   than you want.

achimnol commented:
Renamed the new API to loopUntilNo**More**Events, and rebased the history.
